horovod/spark/common/store.py

```diff
@@ -0,0 +1,334 @@
+    as iterative training using the same DataFrame on different model variations.
+    """
+    def __init__(self):
+        self._train_data_to_key = {}
+        self._val_data_to_key = {}
+
+    def get_filesystem(self):
+        raise NotImplementedError()
+
+    def get_train_data_path(self, idx=None):
+        raise NotImplementedError()
+
+    def get_val_data_path(self, idx=None):
+        raise NotImplementedError()
+
+    def get_test_data_path(self, idx=None):
+        raise NotImplementedError()
+
+    def saving_runs(self):
+        raise NotImplementedError()
+
+    def get_runs_path(self):
+        raise NotImplementedError()
+
+    def get_run_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_checkpoint_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_logs_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_checkpoint_filename(self):
+        raise NotImplementedError()
+
+    def get_logs_subdir(self):
+        raise NotImplementedError()
+
+    def exists(self, path):
+        raise NotImplementedError()
+
+    def read(self, path):
+        raise NotImplementedError()
+
+    def get_full_path_fn(self):
+        raise NotImplementedError()
+
+    def get_local_output_dir_fn(self, run_id):
+        raise NotImplementedError()
+
+    def sync_fn(self, run_id):
+        raise NotImplementedError()
+
+    def to_remote(self, run_id, dataset_idx):
+        attrs = self._remote_attrs(run_id, dataset_idx)
+
+        class RemoteStore(object):
+            def __init__(self):
+                for name, attr in attrs.items():
+                    setattr(self, name, attr)
+
+        return RemoteStore()
+
+    def _remote_attrs(self, run_id, dataset_idx):
+        return {
+            'train_data_path': self.get_train_data_path(dataset_idx),
+            'val_data_path': self.get_val_data_path(dataset_idx),
+            'test_data_path': self.get_test_data_path(dataset_idx),
+            'saving_runs': self.saving_runs(),
+            'runs_path': self.get_runs_path(),
+            'run_path': self.get_run_path(run_id),
+            'checkpoint_path': self.get_checkpoint_path(run_id),
+            'logs_path': self.get_logs_path(run_id),
+            'checkpoint_filename': self.get_checkpoint_filename(),
+            'logs_subdir': self.get_logs_subdir(),
+            'get_full_path': self.get_full_path_fn(),
+            'get_local_output_dir': self.get_local_output_dir_fn(run_id),
+            'sync': self.sync_fn(run_id)
+        }
+
+    @staticmethod
+    def create(prefix_path):
+        if HDFSStore.matches(prefix_path):
+            return HDFSStore(prefix_path)
+        else:
+            return LocalStore(prefix_path)
+
+
+class PrefixStore(Store):
+    def __init__(self, prefix_path, train_path=None, val_path=None, test_path=None, runs_path=None, save_runs=True):
+        self.prefix_path = self.get_normalized_path(prefix_path)
```

Comment:
why should we nomalize the `prefix_path`? Given `prefix_path` as 'hdfs://horovod:8888/path', after normalized the `prefix_path` should be `horovod:8888/path`. Then `train_path` will be 'horovod:8888/path/train_data'. Should this as expect? I guess this path should not be work for spark.write.

Reply:
I see, should wrap with `get_full_path `.

Reply:
You're correct, this should be using the full path.  I've refactored Store to work with full paths by default, and only use the shortened localized paths in specific situations where it's needed (like interacting with a Pyarrow Filesystem object).